Allow the creation and usage of expiryless ProvisionTokens

strideynet opened the pull request:
Closes #20977

Allows `tctl create`/`tctl update` to create tokens that have no expiry from a resource YAML.. 

For now, tokens created using `tctl tokens add` / `tctl nodes add` will still have a default expiry of half an hour applied.

This also sets the groundwork for https://github.com/gravitational/teleport/issues/3327 as well which will allow the creation of tokens w/o expiry using `tctl`.

Docs changes: This will be a separate PR so it only goes into master, so docs show old configuration advice in case customers are not running the latest minor/patch releases.

As part of the TODOs I have investigated all call sites for GenerateToken/UpdateToken/CreateToken RPCs :)

TODO:

- [x] Investigate how this effects `GenerateToken` 
  - GenerateToken already includes a default TTL of 30m.
- [x] Check for other calls of token creation and update to ensure they provide a default and aren't reyling on this default
  - WebUI calls use `/v1/webapi/token` which uses `CreateToken` underneath. `/v1/webapi/token` automatically applies a default TTL of 30 minutes and hence matches the behaviour of `GenerateToken`
 
---

```
➜  teleport git:(strideynet/expiryless-tokens) ✗ build/tctl -c ~/teleports/root/teleport.yaml create -f ./scratch/no-expiry-ghes.yaml
➜  teleport git:(strideynet/expiryless-tokens) ✗ build/tctl -c ~/teleports/root/teleport.yaml get tokens/no-expires-github
kind: token
metadata:
  id: 1675422721279076000
  name: no-expires-github
spec:
  bot_name: github-demo
  github:
    allow:
    - repository: strideynet/sandbox
      repository_owner: strideynet
  join_method: github
  roles:
  - Bot
version: v2
➜  teleport git:(strideynet/expiryless-tokens) ✗ build/tctl -c ~/teleports/root/teleport.yaml tokens ls
Token             Type Labels Expiry Time (UTC)                              
----------------- ---- ------ ---------------------------------------------- 
no-expires-github Bot         never                                          
circleci-bot      Bot         01 Jan 00 00:00 UTC (2562047h47m16.854775807s) 
github-bot        Bot         01 Jan 00 00:00 UTC (2562047h47m16.854775807s) 
```

---

## Release Notes

**Token creation using `tctl create` behavior change**

The default 30 minute expiry will no longer be applied to tokens created using `tctl create` and YAML resource files. 

**CreateToken and UpsertToken RPC deprecation**

The `CreateToken` and `UpsertToken` RPCs are now deprecated, `CreateTokenV2` and `UpsertTokenV2` have been introduced as their successors. Tokens created using `CreateTokenV2` and `UpsertTokenV2` will no longer automatically have a default expiry time applied, if you wish for your tokens to expire, ensure you are setting an expiry.

If you are using the `CreateToken` and `UpsertToken` methods of the Go Teleport API client, in Teleport 13 these have switched to using the new RPCs.

In Teleport 14.0.0, the old RPCs will be removed.

strideynet committed:
Allow the creation and usage of expiryless tokens

strideynet committed:
Tidy up loose ends

strideynet committed:
Fix imports

strideynet committed:
Merge branch 'master' into strideynet/expiryless-tokens

nklaassen commented on lib/services/local/provisioning.go:
```diff
@@ -71,9 +69,6 @@ func (s *ProvisioningService) tokenToItem(p types.ProvisionToken) (*backend.Item
 	if err := p.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-	if p.Expiry().IsZero() || p.Expiry().Sub(s.Clock().Now().UTC()) < time.Second {
-		p.SetExpiry(s.Clock().Now().UTC().Add(defaults.ProvisioningTokenTTL))
-	}
```
This will be a breaking change for any third party clients of the `UpsertToken` api relying on the default expiry. `UpsertToken` is is a part of `api/client` and token creation seems likely to have been automated by some users. It would be unfortunate to have your ephemeral token script suddenly start generating tokens that stick around forever.

We could ship this in teleport 13 with a release note, or add a new `expiryless` (or similar) flag to `ProvisionToken` so that the default behaviour doesn't change, or accept this as a breaking change and move on. Not sure where I stand on that but I think it's worth making a conscious decision.

strideynet replied:
That's true - I hadn't thought RE: external automations people had configured.

My gut feeling is to ship this in a major release and then just announce it as loud as we can as part of upgrade notes.

I think adding an addition field to ProvisionToken feels like it makes it more unwieldy, and just adds additional logic across the codebase we need to account for when displaying these values and when processing tokens on join.

nklaassen reviewed (commented):
could we add some test that default behaviour for existing APIs is unchanged, and that expiryless tokens are allowed?

strideynet commented:
> could we add some test that default behaviour for existing APIs is unchanged, and that expiryless tokens are allowed?

Sure - I'd hope there was already some tests for `GenerateToken` but I'll double check, and see if I can add some elsewhere, as well as adding a case to positively check that an expiryless token can be created :)

strideynet committed:
Merge branch 'master' into strideynet/expiryless-tokens

strideynet committed:
Merge branch 'master' into strideynet/expiryless-tokens

strideynet committed:
Add tests for changes to expiryless behaviour

strideynet committed:
Merge branch 'master' into strideynet/expiryless-tokens

strideynet commented:
Added tests to ensure that default TTL is applied, and that its possible to create a token w/o TTL. That entire suite needs refactoring, but I'll do that as part of deprecating `GenerateToken()` rather than including that in this.

strideynet committed:
Merge branch 'master' into strideynet/expiryless-tokens

strideynet committed:
Use new RPCs for new expiryless behaviour

strideynet committed:
Merge branch 'strideynet/expiryless-tokens' of github.com:gravitational/teleport into strideynet/expiryless-tokens

strideynet committed:
Add tests for new RPCs

strideynet committed:
Merge branch 'master' into strideynet/expiryless-tokens

strideynet commented:
I've introduced a new set of RPCs (CreateTokenV2 and UpsertTokenV2) that will give the new behaviour - this accounts for customers w/ automations directly invoking the RPCs. As part of this work, I've introduced a message for each, rather than placing the ProvisionTokenV2 message directly at the root. This makes it easier to introduce and handle new ProvisionToken versions in future.

strideynet committed:
Remove unused import

strideynet committed:
Recreate behaviour for tokens expiring in past

strideynet committed:
Missing NoError check int ests

strideynet committed:
Merge branch 'master' into strideynet/expiryless-tokens

strideynet committed:
Correctly americanize behavior

strideynet committed:
t push
Merge branch 'strideynet/expiryless-tokens' of github.com:gravitational/teleport into strideynet/expiryless-tokens

hugoShaka reviewed (approved):
Now that the PR is no longer a breaking change, do we want to backport it?

strideynet commented:
> Now that the PR is no longer a breaking change, do we want to backport it?

There's still a change in behaviour here for `tctl create -f` - so I think its worth waiting for a major version so we can communicate this properly.
